---
title: "Robinning"
subtitle: "2- bin quality and taxonomy"
author: "Robin Rohwer"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{css, echo=FALSE}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```

## Bin processing Overview

I saved the output from the big 15,000+ node hour jobs exactly as generated in several places (hard drive, corral, yggshare), but the format I chose for naming the bins in my binning scripts is not ideal for working with the bins. So the first step here is renaming the bins and scaffolds. After that, I used checkM2 to evaluate quality and GTDB-tk to assign taxonomy. I also consider removing short contigs but decide against it for now. 

## Renaming

### bins

The naming structure of the bins straight out of my binning workflow was:  

folder name | file names
:--------|:-----------
assembly it came from | numbered bins
ex: |  
ME2000-03-15pf_3300044539 | bin.1.fa<br>bin.2.fa<br>bin.3.fa<br>...<br>bin.141.fa  

So I renamed them to still be nested in by-assembly folders but now follow the structure  
`samplename_binninggroup_binnumber.fna`.  
Note that samplename is the `sample name _ the JGI taxon OID`, and that the file extension changed to `.fna` to play nicer with checkM2.  

I created a table of all the bin names and what samples they're in:

```{r, eval=TRUE, echo=T, collapse=TRUE}
x <- readRDS("../data/2022-09-14_list_of_all_bins/bin_key.rds")
head(x)
```

The script that did the renaming was generated in `TYMEFLIES/server-scripts/written/2022-09-14_rename_bins` and looks like
```{bash, eval=FALSE}
#!/bin/bash
cd ME2000-03-15pf_3300044539 ; mv bin.100.fa ME2000-03-15pf_3300044539_group1_bin100.fna ; cd ../
cd ME2000-03-15pf_3300044539 ; mv bin.101.fa ME2000-03-15pf_3300044539_group1_bin101.fna ; cd ../
cd ME2000-03-15pf_3300044539 ; mv bin.102.fa ME2000-03-15pf_3300044539_group1_bin102.fna ; cd ../
cd ME2000-03-15pf_3300044539 ; mv bin.103.fa ME2000-03-15pf_3300044539_group1_bin103.fna ; cd ../
```

### scaffolds

Next I renamed the scaffolds, so that when I work with them I know which bin it came from. This is in `TYMEFLIES/server-scripts/written/2022-10-16_rename_bin_scaffolds`.

slurm:
```{bash, eval=F}
#!/bin/bash
#SBATCH -J renaming
#SBATCH -p normal
#SBATCH -N 1
#SBATCH -n 1
#SBATCH -t 48:0:00
#SBATCH -o renaming.o%j
#SBATCH -e renaming.e%j
#SBATCH --mail-type=all
#SBATCH --mail-user=rohwer@utexas.edu
module load launcher
export LAUNCHER_JOB_FILE=rename_scaffolds.launcher
export LAUNCHER_SCHED=interleaved
${LAUNCHER_DIR}/paramrun
```
launcher:
```{bash, eval=F}
module load Rstats ; Rscript rename_scaffolds.R metabat2_bins metabat2_bins_renamed
```

## CheckM2

I used CheckM2 instead of CheckM1 because the paper for CheckM2 (with many shared authors with 1) said that it performed better on streamlined genomes, and specifically used Nanopelagibacter as an example of that. CheckM2 uses a machine learning algorithm to assess completeness and considers size, protein calling, and other genome signatures. CheckM1 uses a list of conserved genes for that lineage. Also, CheckM2 is easier to use with many thousands of bins, and is a little faster.

These scripts were generated by `/Users/rohwer/Desktop/TYMEFLIES/server-scripts/written/2022-09-15_run_checkm2/run_checkM_scripts.R`. It took 45 minutes to run. 

slurm:
```{bash, eval=FALSE}
#!/bin/bash
#SBATCH -J checkm2
#SBATCH -p normal
#SBATCH -N 20
#SBATCH -n 40
#SBATCH -t 4:0:00
#SBATCH -o checkm.o%j
#SBATCH -e checkm.e%j
#SBATCH --mail-type=all
#SBATCH --mail-user=rohwer@utexas.edu
module load launcher
export LAUNCHER_JOB_FILE=checkm2.launcher
export LAUNCHER_SCHED=interleaved
${LAUNCHER_DIR}/paramrun
```
launcher:
```{bash, eval=FALSE}
mkdir checkm2_output/ME2000-03-15pf_3300044539 ; module unload python3 ; source /work/08922/rrohwer/ls6/miniforge3/etc/profile.d/conda.sh ; conda activate checkm2 ; checkm2 predict --threads 64 --input metabat2_bins/ME2000-03-15pf_3300044539 --output-directory checkm2_output/ME2000-03-15pf_3300044539
mkdir checkm2_output/ME2000-03-30D8pf_3300042899 ; module unload python3 ; source /work/08922/rrohwer/ls6/miniforge3/etc/profile.d/conda.sh ; conda activate checkm2 ; checkm2 predict --threads 64 --input metabat2_bins/ME2000-03-30D8pf_3300042899 --output-directory checkm2_output/ME2000-03-30D8pf_3300042899
mkdir checkm2_output/ME2000-05-11D8pf_3300042483 ; module unload python3 ; source /work/08922/rrohwer/ls6/miniforge3/etc/profile.d/conda.sh ; conda activate checkm2 ; checkm2 predict --threads 64 --input metabat2_bins/ME2000-05-11D8pf_3300042483 --output-directory checkm2_output/ME2000-05-11D8pf_3300042483
mkdir checkm2_output/ME2000-05-25pf_3300042154 ; module unload python3 ; source /work/08922/rrohwer/ls6/miniforge3/etc/profile.d/conda.sh ; conda activate checkm2 ; checkm2 predict --threads 64 --input metabat2_bins/ME2000-05-25pf_3300042154 --output-directory checkm2_output/ME2000-05-25pf_3300042154
mkdir checkm2_output/ME2000-06-06pf_3300042549 ; module unload python3 ; source /work/08922/rrohwer/ls6/miniforge3/etc/profile.d/conda.sh ; conda activate checkm2 ; checkm2 predict --threads 64 --input metabat2_bins/ME2000-06-06pf_3300042549 --output-directory checkm2_output/ME2000-06-06pf_3300042549
```

And the results of checkM2 were processed with scripts generated by `/Users/rohwer/Desktop/TYMEFLIES/server-scripts/written/2022-09-15_run_checkm2/save_checkM_output.R`.

1-rename_checkm2_output.sh snippet:
```{bash, eval=FALSE}
cd ME2000-03-15pf_3300044539
mv protein_files ME2000-03-15pf_3300044539
mv diamond_output/DIAMOND_RESULTS.tsv diamond_output/ME2000-03-15pf_3300044539.tsv
mv quality_report.tsv ME2000-03-15pf_3300044539.tsv
cd ../
```
2-zip_checkm2_output.sh snippet:
```{bash, eval=FALSE}
cd ME2000-03-15pf_3300044539
tar cvfz ME2000-03-15pf_3300044539.tar.gz ME2000-03-15pf_3300044539
cd ../
```
3-organize_checkm2_output.sh snippet:
```{bash, eval=FALSE}
mkdir checkm2_protein_files
mkdir checkm2_diamond_files
mkdir checkm2_quality_files

cd ME2000-03-15pf_3300044539
mv ME2000-03-15pf_3300044539.tar.gz ../checkm2_protein_files
mv diamond_output/ME2000-03-15pf_3300044539.tsv ../checkm2_diamond_files
mv ME2000-03-15pf_3300044539.tsv ../checkm2_quality_files
cd ../
```

The CheckM2 results were processed into a single file by `TYMEFLIES/scripts/2022-09-17_look_at_all_checkm2_results.R` 
which looks like:
```{r, eval=TRUE, echo=TRUE, collapse=TRUE, max.width='100%'}
x <- readRDS("../data/2022-09-17_checkm2_output/combined_checkm2_quality_files.rds")
head(x)
```

## GTDB-tk

I used GTDB-tk to assign taxonomy to all of the bins. The scripts were generated by `TYMEFLIES/server-scripts/written/2022-09-22_run_gtdb-tk/run_gtdbtk_scripts.R`. I had to be careful about memory, and limit the CPUs used to 20 to avoid errors, so it took 17 hours to run.  

slurm:
```{bash, eval=FALSE}
#!/bin/bash
#SBATCH -J gtdbtk
#SBATCH -p normal
#SBATCH -N 20
#SBATCH -n 20
#SBATCH -t 48:0:00
#SBATCH -o gtdbtk.o%j
#SBATCH -e gtdbtk.e%j
#SBATCH --mail-type=all
#SBATCH --mail-user=rohwer@utexas.edu
module load launcher
export LAUNCHER_JOB_FILE=gtdbtk.launcher
export LAUNCHER_SCHED=interleaved
${LAUNCHER_DIR}/paramrun
```
launcher:
```{bash, eval=FALSE}
mkdir gtdbtk_output/ME2000-03-15pf_3300044539 ; source /work/08922/rrohwer/ls6/miniforge3/etc/profile.d/conda.sh ; conda activate gtdbtk-2.1.1 ; gtdbtk classify_wf --genome_dir metabat2_bins/ME2000-03-15pf_3300044539 --out_dir gtdbtk_output/ME2000-03-15pf_3300044539 --cpus 15 --pplacer_cpus 3 
mkdir gtdbtk_output/ME2000-03-30D8pf_3300042899 ; source /work/08922/rrohwer/ls6/miniforge3/etc/profile.d/conda.sh ; conda activate gtdbtk-2.1.1 ; gtdbtk classify_wf --genome_dir metabat2_bins/ME2000-03-30D8pf_3300042899 --out_dir gtdbtk_output/ME2000-03-30D8pf_3300042899 --cpus 15 --pplacer_cpus 3 
mkdir gtdbtk_output/ME2000-05-11D8pf_3300042483 ; source /work/08922/rrohwer/ls6/miniforge3/etc/profile.d/conda.sh ; conda activate gtdbtk-2.1.1 ; gtdbtk classify_wf --genome_dir metabat2_bins/ME2000-05-11D8pf_3300042483 --out_dir gtdbtk_output/ME2000-05-11D8pf_3300042483 --cpus 15 --pplacer_cpus 3 
mkdir gtdbtk_output/ME2000-05-25pf_3300042154 ; source /work/08922/rrohwer/ls6/miniforge3/etc/profile.d/conda.sh ; conda activate gtdbtk-2.1.1 ; gtdbtk classify_wf --genome_dir metabat2_bins/ME2000-05-25pf_3300042154 --out_dir gtdbtk_output/ME2000-05-25pf_3300042154 --cpus 15 --pplacer_cpus 3 
mkdir gtdbtk_output/ME2000-06-06pf_3300042549 ; source /work/08922/rrohwer/ls6/miniforge3/etc/profile.d/conda.sh ; conda activate gtdbtk-2.1.1 ; gtdbtk classify_wf --genome_dir metabat2_bins/ME2000-06-06pf_3300042549 --out_dir gtdbtk_output/ME2000-06-06pf_3300042549 --cpus 15 --pplacer_cpus 3 
```

I also generated scrips to check for errors and format the output using `TYMEFLIES/server-scripts/written/2022-09-22_run_gtdb-tk/save_gtdbtk_output.R`

1-check_for_gtdbtk_errors.sh snippet:
```{bash, eval=FALSE}
#!/bin/bash
echo gtdbtk error outputs > all_errors.txt


cd ME2000-03-15pf_3300044539
echo ME2000-03-15pf_3300044539 >> ../all_errors.txt
cat gtdbtk.warnings.log >> ../all_errors.txt
cd ../
```
2-put_all_output_summaries_into_single_folder.sh snippet:
```{bash, eval=FALSE}
#!/bin/bash

cd ME2000-03-15pf_3300044539
cp gtdbtk.bac120.summary.tsv ME2000-03-15pf_3300044539.tsv
mv ME2000-03-15pf_3300044539.tsv ../gtdbtk_bac120_summary_files/
cd ../
```

And I processed the results into a single file using `2022-09-25_1-look_at_all_the_gtdbtk_output.R` which created this file:
```{r, eval=TRUE, echo=TRUE, collapse=TRUE, max.width='100%'}
x <- readRDS("../data/2022-09-22_gtdb-tk_output/bin_taxonomy.rds")
head(x)
```

## Exploring results

I combined the bin quality and taxonomy files using `2022-09-25_2-combine_checkm_and_gtdb_info.R`:
```{r, eval=TRUE, echo=TRUE, collapse=TRUE, max.width='100%'}
x <- readRDS("../data/2022-09-23_bin_stats/all_bins.rds")
head(x)
```
And I looked at the results using `2022-09-25_3-explore_whats_in_the_bins.R` which made plots saved in `figures/2022-09-25_explore_the_bins`. Orange is MQ (completeness >= 50, contamination <10), Red is HQ (completeness >= 90, contamination <50). Note though that overall contamination is super low, the vast majority of all bins are under 5!

![](../figures/2022-09-25_explore_the_bins/png_for_markdown/0_all_All Bins-table.png){width=100%}

Nanopelagicales also look good, lots of bins, an issue with completeness though:

![](../figures/2022-09-25_explore_the_bins/png_for_markdown/7_order_o__Nanopelagicales-table.png){width=100%}

And specifically can see that both acI-A and acI-B are well represented:

![](../figures/2022-09-25_explore_the_bins/7_order_o__Nanopelagicales-9.png){width=50%, height=50%}

![](../figures/2022-09-25_explore_the_bins/7_order_o__Nanopelagicales-10.png){width=50%, height=50%}
You can look at whatever lineage with the R script, FYI.


